### ISSUE TITLE
An ability to reload full construction on poke call

### ISSUE THREAD

**Issue by IrwinBridge** [grych/drab#71]
Maybe some possibility to mark a construction somehow to know Drab that it should reload only this block.

**Comment by grych** [comment]
This is an original post from [exlirforum.com](https://elixirforum.com/t/how-to-enforce-drab-to-reload-full-loop/11262):

---

I use Drab and I want just to change the value of assign inside of a loop.

I have this loop code:

```eex
<%= for page <- @holder.pages do %>
  <%= if page.id == @page_id do %>
    <tr>
      <td><%= page.title %></td>
    </tr>
  <% end %>
<% end %>
```
I change @page_id like this:

Drab.Live.poke(socket, page_id: id)

This function called when clicking on a button.
Everything works just fine when I place @page_id on some static element like h1.
But it doesn't work inside of the loop.
Error: undefined function page/0

Can someone explain me why does it happen? I think Drab doesn't fully reload loop - only if expression.

**Comment by grych** [comment]
Unfortunately, it is not a bug. This is a limitation of Drab.Live's local variables scope && the way how Drab is replacing parts of the page with the evaluated expression with new assign values.

When you're poking the new version of assign @page_id, Drab tries to update it in every expression on a page, which contains that value. In this case, it is if page.id == @page_id, do:.... But in the particular moment it does not know the value of page variable - that's why it shows undefined function page() while evaluating this expression.

It is done by purpose. I could search expressions for assigns in their do..end blocks, and re-render the the whole for comprehension while changing @page_id. And actually it was made like this before, but I realized that in the real world it provides to full reload of almost the whole page and more issues, like #42.

In this particular situation, when you poke @page_id, you really want to update the whole for compehension. One of the solution would be to put @page_id assign in the comprehension filter:

```eex
<%= for page <- @holder.pages, page.id == @page_id do %>
  <tr>
    <td><%= page.title %></td>
  </tr>
<% end %>
```

This moves @page_id to the for expression, and now, after you poke it, it re-evaluates the whole one.

I know it is not very intuitive, but so far I couldn't find the better solution.

Docs: https://hexdocs.pm/drab/Drab.Live.EExEngine.html#module-limitations2

**Comment by grych** [comment]
TODO: solve such issues with updating both assigns in the same poke:

    poke socket, page: peek(:page), page_id: new_page_id

(or even introduce new API)

    poke socket, page_id: new_page_id, refresh_assigns: [:page]

In the current version poking both assigns will generate the same error, because Drab will try to update `@page_id` anyway, even if it is completely pointless as it is inside the `for` comprehension, which is updated anyway.

Drab should be aware of such cases and update only the parent statement. It would be also a performance improvement.

**Comment by grych** [comment]
Drab now is able to detect this case. Change will be included in 0.7.1. 

Solution for your case:

    poke socket, page: peek(:page), page_id: new_page_id

I did not introduce a new API for `poke`, as I don't want to add a special options like `refresh_assigns`. You may have an assign named `@refresh_assigns` in you page.

If you have an idea for a good API for this case, please do not hesitate.
